man/man1/ggobi.1
.TH ggobi 1 "30 September 2000"
.SH NAME
.B ggobi 
\- interactive dynamic graphics program for data visualization

.SH SYNOPSIS
.B ggobi
[
.B \-xml
]
[
.B \-ascii
]
[
.B \-version
.B \--version
]
filename

.SH DESCRIPTION
.I ggobi
is an interactive dynamic graphics program for data
visualization, especially designed for the exploration of
multivariate data.  It represents a significant redesign of 
its precedessor
.I xgobi (www.research.att.com/areas/stat/xgobi),
with multiple plotting windows, a color lookup table manager, an XML
file format for data, and other changes.  It can be embedded into
other software (such as R, Perl, Python,...) and controlled using an
API (Application Programming Interface).

GGobi's plot types include scatterplots, scatterplot matrices,
parallel coordinates plots, and time series plots.  These are some of
the tools available for plot display and manipulation:
.RS .25i
.IP o 2
Cycling rapidly through two-variable scatterplots.
.IP o 2
Grand tours and correlation tours: smooth randomized sequences of
two-dimensional projections in order to explore a higher-dimensional
point cloud of multivariate data.  Projections can be manually
controlled and optimized with projection pursuit.
.IP o 2
Brushing: changing the color or plotting
character of a point or a group of points.
.IP o 2
Hiding groups:  using checkboxes to specify which brushed
groups should be plotted, or included in calculations.
.IP o 2
Scaling: moving and reshaping a plot interactively.
.IP o 2
Identification: displaying a label next to a point in a scatterplot
or scatterplot matrix.
.IP o 2
Linked views: Brushing, identification and touring are linked;
that is, actions in one window are immediately reflected in
other windows.
.IP o 2
Moving points: Points can be moved in data space by dragging them
with the mouse on the screen.
.IP o 2
Subsetting: Subsamples can be drawn either systematically or randomly,
which is particularly helpful for large datasets.
.IP o 2
Jittering: Variables can be jittered, which is useful for 
visually spreading out tied data values, in particular
the values of discrete variables.
.IP o 2
Missing values: are accepted and can be dealt with by imputation of
constant values, random values, or user-supplied imputed values.
Missing value patterns can be examined in a separate linked ggobi
window.
.IP o 2
Variable transformations: A menu of transformations such as Box-Cox is available.

''' .IP o 2
''' Smooths: Smooth curves can be drawn through point clouds.
''' .IP o 2
''' Case label list: shown in a separate window, linked to the main
''' window for labeling points.
''' .IP o 2
''' Postscript plotting: The contents of the plot windows can be written
''' out to postscript files for high-quality plotting.
.RE

GGobi has a direct manipulation interface, and all the above actions
are performed using the mouse.

In addition, ggobi can be treated as a C library and directly
embedded in (or linked with) other software.  For example, it can be
controlled from the R command line or from a Perl or Python script.

Execute
.IR help(ggobi)
within R for information; if the ggobi function
has not been installed, see www.ggobi.org for instructions.

.SH OPTIONS

.B \-xml
.LP
.RS .25i
Use a single XML file as input.  The XML format is documented in 
XML.pdf, which is included in the ggobi distribution.  See the
sample data files.
.RE

.B \-ascii
.LP
.RS .25i
Use a set of ASCII files as input.  See below for more a detailed
description.  These files are almost compatible
with the files used by xgobi, but the .color files use numbers between
1 and 10 instead of color names, and some files are not accepted
(.erase, .vgroups, .edges, .nlinkable, .rgroups).  
.RE

.SH FILES
ggobi's file input can take two basic forms:  (1) the style
of multiple files, sharing a common basename, that was used
in xgobi, and (2) one file in XML.  A utility, xmlConvert,
is provided to transform the multiple-file format into XML format.

The XML format allows multiple datasets to be included within
a single file.  Those datasets can be related using the
.I id
tag.  If cases in two datasets have the same id, they are
linked for brushing and identification.  The id tag can also
be used to specify the endpoints for a line segment, or edge,
between two points.  See XML.pdf for a more detailed explanation.

If the file type is not specified on the command line, ggobi
will attempt to identify whether the file is XML or ASCII, and
if the suffix hasn't been specified, it will first look for an XML
version of the data.

The multiple-file format is described here. 

.nf
.B filename.dat
.fi
.RS .25i

The data input file
should be an ASCII file with the data matrix arranged in rows and
columns; in ASCII, rows must be distinguished by carriage
returns, and columns can be separated by any amount of white
space.  ("Row" is synonymous with "case", "record"; 
"column" is synonymous with "variable", "attribute".)
Missing values can be coded as ".", "NA" or "na".  (The
input file can also be a binary file, which can be produced
within ggobi once the ASCII data has been read in.)  

In addition to the data file, ggobi accepts
the following input files for the display of the data:

.nf
.B filename.row
.B filename.rowlab
.B filename.case
.fi
.RS .25i

Row or case labels: a label for each row of the data matrix, which is
displayed in the identification mode.  The file should contain one
label per line.

.RE

.nf
.B filename.col
.B filename.column
.B filename.collab
.B filename.var
.fi
.RS .25i

Column or variable labels: a label for each column of the data matrix,
which becomes part of the ggobi variable selection panel.  The file
should contain 
.I
one label per line.
.R
The label may contain white space.

To specify axis ranges, use the vertical bar "|" as a field separator
and add minimum and maximum values, so that the line looks like
this:  varname|minval|maxval
.RE

.nf
.B filename.colors
.fi
.RS .25i

Brushing colors: an integer for each case, one per line,
representing the index of a vector of colors, in the range [0:ncolors-1].
(Note again that this differs from XGobi's color format, which used
X color names.  ggobi uses "anonymous" color codes because of its
color lookup table manager which allows users to remap colors
arbitrarily.)
.RE

.nf
.B filename.glyphs
.fi
.RS .25i

Brushing glyphs: a glyph type and size for each point in the plot,
representing a row or case of the data.  The glyph types are as follows:
.RS .25i

.nf
plus: the glyph '+'
x: the glyph 'x'
or: an open rectangle
fr: a filled rectangle
oc: an open circle
fc: a filled circle
.: a single-pixel point
.fi

The sizes range from 0 to 7.

Note:  Users of Microsoft Windows should avoid using the circle, since these
are drawn very slowly.


.RE

.nf
.B filename.missing
.fi
.RS .25i

A file identical in structure to filename.dat, where non-zero values
indicate positions with missing (or censored, or otherwise
exceptional) values.  This file represents the pattern of missing
values in the data; it can be examined in a separate window by
selecting
.I New scatterplot 
or
.I New scatterplot matrix
from the
.I MISSING VALUES PLOTS
submenu of the
.I Window
menu.
.RE


.SH AUTHORS
.nf
Deborah Swayne (dfs@research.att.com)
Dianne Cook (dicook@iastate.edu)
Duncan Temple Lang (duncan@research.bell-labs.com)
Andreas Buja (andreas@research.att.com)
.fi
.SH CONTACT
Deborah Swayne (dfs@research.att.com)
.SH SEE ALSO
Sample data files are included in the ggobi distribution.

For papers and further pointers, check www.ggobi.org, and
the xgobi web site, www.research.att.com/areas/stat/xgobi.

In particular, the manual is in manual.pdf.  The XML format
is described in XML.pdf.  The manual for use of R and GGobi
together is RGGobi.pdf.

.SH BUGS
Probably too numerous to mention, but you do the authors a favor if
you report the ones you notice.




NEWS
GGobi build system ported to automake and this file created.


ggobi.c
/*-- ggobi.c --*/
/*
    This software may only be used by you under license from AT&T Corp.
    ("AT&T").  A copy of AT&T's Source Code Agreement is available at
    AT&T's Internet website having the URL:
    <http://www.research.att.com/areas/stat/ggobi/license.html>
    If you received this software without first entering into a license
    with AT&T, you have an infringing copy of this software and cannot use
    it without violating AT&T's intellectual property rights.
*/
#define GGOBIINTERN

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <gtk/gtk.h>

#include "ggobi.h"

#include "vars.h"
#include "externs.h"

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "print.h"

#ifdef USE_XML
#include "read_init.h"
#endif
#include "colorscheme.h"

#ifdef WIN32
#define DIR_SEPARATOR '\\'
#else
#define DIR_SEPARATOR '/'
#endif

static GGobiOptions sessionoptions;
GGobiOptions *sessionOptions;


ggobid **all_ggobis;
gint num_ggobis;
gint totalNumGGobis;

   /* Needs to be connected to MAXNVARS in scatmat.c
      and MAXNPCPLOTS in  parcoords.c
    */
#define MAXNVARS 4 
#define MAXNPCPLOTS 5
#define MAXNTSPLOTS 6 


const gchar * const ViewTypes[] =
  {"Scatterplot", "Scatterplot Matrix", "Parallel Coordinates"};
const gint ViewTypeIndices[] = {scatterplot, scatmat, parcoords};           
const gchar *const DataModeNames[num_data_modes] =
  {"ASCII", "binary", "R/S data", "XML", "MySQL", "URL", "Unknown"};


void initSessionOptions();

gchar *
getOptValue(const gchar * const name, const gchar * const value)
{
    const gchar * ptr = (const gchar *) value;

    if(ptr[0] != '-' || ptr[1] != '-')
	return(NULL);

    if(strncmp(name, value + 2, strlen(name)) == 0) {
        ptr = value + strlen(name) + 2;
        if(ptr[0] != '=' || ptr[1] == '\0') {
            g_printerr("--%s must be given a value in the form --%s=value\n", name, name);
            fflush(stderr);          
            ptr = NULL;
	} else
	    ptr = g_strdup(ptr+1);
    } else
	ptr = NULL;

    return((gchar *)ptr);
}

gint
parse_command_line (gint *argc, gchar **av, ggobid *gg)
{
  gboolean stdin_p = false;
  gchar *ptr;

/*
 * Now parse the command line.
*/
  for ( ; *argc>1 && av[1][0]=='-'; (*argc)--,av++) {
    /*
     * -s:  ggobi initiated from inside S
    */
    if (strcmp (av[1], "-s") == 0)
      sessionOptions->data_mode = Sprocess_data;
    else if (strcmp (av[1], "-ascii") == 0) {
      sessionOptions->data_mode = ascii_data;
    }
    else if (strcmp (av[1], "-xml") == 0) {
#ifdef USE_XML
      sessionOptions->data_mode = xml_data;
#else
      g_printerr("No xml support compiled for this version, ignoring %s\n", av[1]);
#endif
    } else if (strcmp (av[1], "-v") == 0 || strcmp (av[1], "--validate") == 0) {
#ifdef USE_XML
      extern gint xmlDoValidityCheckingDefaultValue;
      xmlDoValidityCheckingDefaultValue = 1;
#else
      g_printerr("No xml support compiled for this version, ignoring %s\n", av[1]);
#endif
   }

   else if(strcmp(av[1], "--verbose") == 0  || strcmp(av[1], "-verbose") == 0  || strcmp(av[1], "-V") == 0) {
      sessionOptions->verbose = GGOBI_VERBOSE;
   } 

   else if(strcmp(av[1], "--silent") == 0  || strcmp(av[1], "-silent") == 0)  {
      sessionOptions->verbose = GGOBI_SILENT;
   } 
 #ifdef USE_MYSQL
    else if (strcmp (av[1], "-mysql") == 0) {
    }
#endif
    /*
     * -:  look to stdin for the input data
    */
    else if (strcmp (av[1], "-") == 0) {
      stdin_p = true;
    }

    /*
     * -version:  print version date, return
    */
    else if (strcmp (av[1], "-version") == 0) {
      g_printerr ("This version of GGobi is dated %s\n", GGOBI(getVersionDate()));
      exit (0);

    } else if (strcmp (av[1], "--version") == 0) {
      g_printerr ("%s\n", GGOBI(getVersionString()));
      exit (0);
    } else if(strcmp(av[1], "-init") == 0) {
#ifdef SUPPORT_INIT_FILES
      sessionOptions->initializationFile = g_strdup(av[2]);
#else
      g_printerr ("-init not supported without XML\n");fflush(stderr);
#endif
      (*argc)--; av++;
    } else if((ptr = getOptValue("init", av[1]))) {
#ifdef SUPPORT_INIT_FILES
	    sessionOptions->initializationFile = ptr;
#endif
    } else if(strcmp(av[1], "-noinit") == 0) {
#ifdef USE_XML
      sessionOptions->initializationFile = g_strdup("");
#endif
    } else if(strcmp(av[1],"-colorschemes") == 0) {
#ifdef USE_XML
      sessionOptions->info->colorSchemeFile = av[2];
      /* read_colorscheme(av[2], &(sessionOptions->colorSchemes)); */
#else
      g_printerr ("-colorschemes not supported without XML\n"); fflush(stderr);
#endif
      (*argc)--; av++;
    } else if((ptr = getOptValue("colorschemes", av[1]))) {
#ifdef USE_XML
       sessionOptions->info->colorSchemeFile = ptr;
#else
      g_printerr ("--colorschemes not supported without XML\n"); fflush(stderr);
#endif
    } else if(strcmp(av[1],"-activeColorScheme") == 0) {
#ifdef USE_XML
      sessionOptions->activeColorScheme = g_strdup(av[2]);
#else
      g_printerr ("-activeColorScheme not supported without XML\n"); fflush(stderr);
#endif
      (*argc)--; av++;
    } else if((ptr = getOptValue("activeColorScheme", av[1]))) {
#ifdef USE_XML
	sessionOptions->activeColorScheme = ptr;
#else
      g_printerr ("--activeColorScheme not supported without XML\n"); fflush(stderr);
#endif

    }  else if(strcmp(av[1], "-datamode") == 0) {
      sessionOptions->data_type = g_strdup(av[2]);
      (*argc)--; av++;
    }  else if((ptr = getOptValue("datamode", av[1]))) {
	sessionOptions->data_type = ptr;
    }
  }

  (*argc)--;
  av++;

/*
 * Test the values
*/

  if (*argc == 0)
    sessionOptions->data_in = (stdin_p) ? g_strdup_printf ("stdin") : NULL;
  else
    sessionOptions->data_in = g_strdup(av[0]);

  return 1;
}

gint GGOBI(main)(gint argc, gchar *argv[], gboolean processEvents);

gint 
main(gint argc, gchar *argv[])
{ 
 GGOBI(main)(argc, argv, true);
 return (0);
}

gint
ggobi_remove (ggobid *gg)
{ 
  gint i;
  for (i = 0; i < num_ggobis; i++) {
    if (all_ggobis[i] == gg) {
      return (ggobi_remove_by_index (gg, i));
    }
  }

  return (-1);
}

/*
      Also, need to close and free the displays associated with the ggobi.
 */
gint
ggobi_remove_by_index (ggobid *gg, gint which)
{
  GSList *l;
  datad *d;
  gint numDatasets, i;

  /* Move all the entries after the one being removed
     down by one in the array to compact it.
   */
  if(which < num_ggobis -1) {
     memcpy(all_ggobis + which,
            all_ggobis + which +
            1,
            sizeof(ggobid*)*(num_ggobis-which-1));
  }
  /* Now patch up the array so that it has the correct number of elements. */
  num_ggobis--;
  if (num_ggobis > 0)
    all_ggobis = (ggobid**)
      g_realloc (all_ggobis, sizeof(ggobid*) * num_ggobis);
  else
    all_ggobis = NULL;

  /* 
      This was crashing in R. Probably because when we exhaust the list
      and remove the final element, we get back garbage.
      This isn't a problem in stand-alone as it never gets called.
   */
  numDatasets = g_slist_length (gg->d);
  for (i=0,l = gg->d; l != NULL && i < numDatasets; i++, l = gg->d) {
    d = (datad *) l->data;
    datad_free (d, gg);
    gg->d = g_slist_remove (gg->d, d);
  }

  g_free (gg);

  return (which);
}

/*
 The code within the TEST_KEYS sections performs a test of handling key presses on 
 the numbered keys. It registers a function
 */
#ifdef TEST_KEYS
gboolean
DummyKeyTest(guint keyval, GtkWidget *w, GdkEventKey *event,  cpaneld *cpanel, splotd *sp, ggobid *gg, void *userData)
{
  static gint count = 0;
  fprintf(stderr, "Key press event (count = %d): key = %d, data = %s\n", 
                     count, (gint)keyval, (gchar *)userData);
  fflush(stderr);

  if(++count == 4) {
    count = 0;
    GGOBI(removeNumberedKeyEventHandler)(gg);
  }
 return(true);
}
#endif

/**
  Find the color scheme element in the list with the specified
  name.
 */
colorschemed *
findColorSchemeByName(GList *schemes, const gchar *name)
{
  colorschemed *s;
  gint i, n;

  n = g_list_length(schemes);
  for(i = 0; i < n; i++) {
   s = (colorschemed *)g_list_nth_data(schemes, i);
   if(strcmp(name, s->name) == 0)
       return(s);
  }
  return(NULL);
}

ggobid*
ggobi_alloc()
{
 ggobid *tmp;

  tmp = (ggobid*) g_malloc (sizeof (ggobid));

  memset (tmp, '\0', sizeof (ggobid));
  tmp->firsttime = true;
  tmp->brush.firsttime = true;

  /*-- initialize to NULLMODE and check for ncols later --*/
  tmp->viewmode = NULLMODE;
  tmp->prev_viewmode = NULLMODE;
  tmp->projection = NULLMODE;
  tmp->prev_projection = NULLMODE;
  /*-- --*/

  /*-- initialize main window, tool windows to NULL --*/
  tmp->main_window = NULL;
  tmp->display_tree.window = NULL;
  tmp->vartable_ui.window = NULL;
  tmp->sphere_ui.window = NULL;
  tmp->cluster_ui.window = NULL;
  tmp->color_ui.symbol_window = NULL;
  /*-- --*/

  tmp->color_ui.margin = 10;
  tmp->tour2d.idled = 0;
  tmp->tour1d.idled = 0;
  tmp->tourcorr.idled = 0;
  tmp->tour1d.fade_vars = true;
  tmp->tour2d.fade_vars = true;
  tmp->tourcorr.fade_vars = true;
  tmp->brush.updateAlways_p = true;

  tmp->printOptions = NULL;
  tmp->pluginInstances = NULL;

  tmp->colorSchemes = sessionOptions->colorSchemes;
  if (sessionOptions->activeColorScheme)
    tmp->activeColorScheme = findColorSchemeByName(tmp->colorSchemes,
      sessionOptions->activeColorScheme);
  else {
    /*-- use "Spectrum 7" by default, if it's present --*/
    sessionOptions->activeColorScheme = "Spectrum 7";
    tmp->activeColorScheme = findColorSchemeByName(tmp->colorSchemes,
      sessionOptions->activeColorScheme);
    if (!tmp->activeColorScheme)
      tmp->activeColorScheme = (colorschemed *)
        g_list_nth_data(tmp->colorSchemes, 0);
  }
  colorscheme_init (tmp->activeColorScheme);
  

  totalNumGGobis++;

  all_ggobis = (ggobid **)
    g_realloc (all_ggobis, sizeof(ggobid*)*(num_ggobis+1));
  all_ggobis[num_ggobis] = tmp;
  num_ggobis++;

#ifdef TEST_KEYS
  GGOBI(registerNumberedKeyEventHandler)(DummyKeyTest,
    g_strdup("A string for the key handler"),"Test handler", NULL, tmp, C);
#endif

  return (tmp);
}


  /* Available so that we can call this from R
     without any confusion between which main().
   */
gint 
GGOBI(main)(gint argc, gchar *argv[], gboolean processEvents)
{
  extern void make_ggobi (GGobiOptions *, gboolean, ggobid *);
  GdkVisual *vis;
  ggobid *gg;
  initSessionOptions();
  sessionOptions->cmdArgs = argv;
  sessionOptions->numArgs = argc;

  gtk_init (&argc, &argv);

  vis = gdk_visual_get_system ();

  parse_command_line (&argc, argv, gg);


#ifdef SUPPORT_INIT_FILES
  process_initialization_files();
#endif

  if(sessionOptions->verbose == GGOBI_VERBOSE)
    g_printerr("progname = %s\n", g_get_prgname());

  if(sessionOptions->verbose == GGOBI_VERBOSE)
    g_printerr("data_in = %s\n", sessionOptions->data_in);

  if(DefaultPrintHandler.callback == NULL)
    setStandardPrintHandlers();

#ifdef USE_XML
  if(sessionOptions->info->colorSchemeFile && sessionOptions->colorSchemes == NULL) {
      read_colorscheme(sessionOptions->info->colorSchemeFile, &sessionOptions->colorSchemes);
  }
#endif

  if(sessionOptions->colorSchemes == NULL) {
      /* Debby, create the default color scheme from the built-in data 
         and append. */
      colorschemed *scheme = default_scheme_init();
      sessionOptions->colorSchemes = g_list_append(sessionOptions->colorSchemes, scheme);
      sessionOptions->activeColorScheme = scheme->name;
  }
  

  gg = ggobi_alloc();

  gg->mono_p = (vis->depth == 1 ||
               vis->type == GDK_VISUAL_STATIC_GRAY ||
               vis->type == GDK_VISUAL_GRAYSCALE);

  make_ggobi (sessionOptions, processEvents, gg);

  /* g_free (sessionOptions->data_in); */

  return (num_ggobis);
}


void
initSessionOptions()
{
  sessionOptions = &sessionoptions;
  sessionOptions->data_mode = unknown_data;

  sessionOptions->showControlPanel = true;
  sessionOptions->verbose = GGOBI_CHATTY;


  sessionOptions->info = (GGobiInitInfo*) g_malloc(sizeof(GGobiInitInfo));
  memset(sessionOptions->info, '\0', sizeof(GGobiInitInfo));
  sessionOptions->info->glyph.size = sessionOptions->info->glyph.type = -1;
  sessionOptions->info->createInitialScatterPlot = true;
  sessionOptions->info->allowCloseLastDisplay = false;
  sessionOptions->info->quitWithNoGGobi = true; 
  sessionOptions->info->numScatMatrixVars = MAXNVARS;
  sessionOptions->info->numParCoordsVars = MAXNPCPLOTS;
  sessionOptions->info->numTimePlotVars = MAXNTSPLOTS;
#ifdef USE_XML
#endif
}


/*
  Called in response to a window being destroyed.
  Be careful not to use the event or the object
  as they are not guaranteed to be correct. We are 
  abusing the callback marshalling system a little.
 */
gboolean
ggobi_close (ggobid *gg, GdkEvent *ev, GtkObject *w)
{
  GGOBI(close)(gg, true);
  return(true);
}


/*
   Key for storing a reference to a ggobid instance in a widget
   so that we can retrieve it within a callback.
*/
const gchar * const GGobiGTKey = "GGobi";

const gchar* const key_get (void) {
  return GGobiGTKey;
}

/*
  Computes the ggobid pointer associated with the specified
  widget. It does so by looking in the window associated with the widget
  and then looking for an entry in the window's association table.
  This assumes that the ggobid reference was stored in the window 
  when it was created.
 */
ggobid*
GGobiFromWidget (GtkWidget *w, gboolean useWindow)
{
  /*
   GdkWindow *win = gtk_widget_get_parent_window(w);
   return(GGobiFromWindow(win));
  */
  ggobid *gg = NULL;
  gg = (ggobid*) gtk_object_get_data (GTK_OBJECT(w), GGobiGTKey);
  ValidateGGobiRef (gg, true);

  return (gg);
}

ggobid* GGobiFromWindow (GdkWindow *win)
{
  ggobid *gg = NULL;
  gg = (ggobid*) gtk_object_get_data(GTK_OBJECT(win), GGobiGTKey);
  ValidateGGobiRef (gg, true);

  return(gg);
}

ggobid* 
GGobiFromSPlot(splotd *sp)
{
 return (sp->displayptr->ggobi);
}

ggobid* 
GGobiFromDisplay(displayd *display)
{
  return (display->ggobi);
}

void
GGobi_widget_set (GtkWidget *w, ggobid *gg, gboolean asIs)
{
  GtkObject *obj;
  if (asIs)
    obj = GTK_OBJECT (w);
  else 
    obj = GTK_OBJECT (gtk_widget_get_parent_window (w));

  gtk_object_set_data (obj, GGobiGTKey, gg);
}


ggobid *
ggobi_get(gint which)
{
 extern ggobid** all_ggobis;
 if(which > -1 && which < num_ggobis)
   return (all_ggobis[which]);
 else
   return(NULL);
}

gint
ggobi_getIndex(ggobid *gg)
{
  gint i;
  for(i = 0; i < num_ggobis ; i++) {
    if(all_ggobis[i] == gg)
      return(i);
  }

 return(-1);
}

datad *
GGobi_get_data(gint which, const ggobid * const gg)
{
   datad *d;
   d = g_slist_nth_data(gg->d, which);

   return(d);
}

datad *
GGobi_get_data_by_name(const gchar * const name, const ggobid * const gg)
{
   datad *d;
   GSList *l;

   for (l = gg->d; l; l = l->next) {
     d = (datad *) l->data;
     if(strcmp(d->name, name) == 0)
	 return(d);
   }
   return(NULL);
}


ggobid*
ValidateGGobiRef(ggobid *gg, gboolean fatal)
{ 
 extern ggobid** all_ggobis;
 extern gint num_ggobis;
  gint i;
  for(i = 0; i < num_ggobis ; i++) {
   if(all_ggobis[i] == gg)
    return(gg);
  }

  g_printerr ("Incorrect reference to ggobid.\n");

 if (fatal)
  exit(10);

 return(NULL);
}

datad *
ValidateDatadRef(datad *d, ggobid *gg, gboolean fatal)
{
  gint i, n;
  n = g_slist_length(gg->d);
  for(i = 0; i < n ; i++) {
   if(g_slist_nth_data(gg->d, i) == d)
    return(d);
  }

  g_printerr("Incorrect reference to datad.\n");

 if (fatal)
  exit(11);

 return(NULL); 
}



displayd *
ValidateDisplayRef(displayd *d, ggobid *gg, gboolean fatal)
{
  gint i, n;
  n = g_list_length(gg->displays);
  for(i = 0; i < n ; i++) {
   if(g_list_nth_data(gg->displays, i) == d)
    return(d);
  }

  g_printerr("Incorrect reference to display.\n");

 if (fatal)
  exit(11);

 return(NULL); 
}

#ifdef USE_XML
/*
  Determines which initialization file to use
  Checks for the one specified by
    1) the -init command line option
    2) the GGOBIRC environment variable
    3) the $HOME/.ggobirc file.
    4) ggobirc in the directory of the executable (argv[0]) 
 */
void
process_initialization_files()
{
  GGobiInitInfo *info;
  gchar buf[100];
  const gchar *fileName = NULL;

  if(sessionOptions->initializationFile)
    fileName = sessionOptions->initializationFile;
  else {
           /* This use of getenv() is not portable. */
    fileName = getenv("GGOBIRC");
    if(!fileName || !fileName[0]) {
      gchar *tmp;
      /* HOME doesn't necessarily make sense in Windows. */
      tmp = getenv("HOME");
      if(tmp) {
        sprintf(buf, "%s/.ggobirc", tmp);
        fileName = buf;
      } else {
        gchar *v;
        tmp = g_strdup(sessionOptions->cmdArgs[0]);
        v = strrchr(tmp, DIR_SEPARATOR);
        if(v) {
	  v[1] = '\0';
	}
        sprintf(buf, "%sggobirc",tmp);
        fileName = buf;
        g_free(tmp);
      }
    }
    if(fileName)
      sessionOptions->initializationFile = g_strdup(fileName);
  }
     
  if(fileName && fileName[0] && canRead(fileName)) {
    info = read_init_file(fileName, sessionOptions->info);
    /* sessionOptions->info = info; */
  }
}
#endif
